org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/FilesUtils.java

```diff
@@ -145,4 +152,153 @@ static String toString(InputStream is) {
 			return s.hasNext() ? s.next() : "";
 		}
 	}
+
+	public static String getPathFromURI(String uri) {
+		int offsetAfterScheme = getOffsetAfterScheme(uri);
+		if (uri == null) {
+			return null;
+		}
+
+		int uriLength = uri.length();
+		if (offsetAfterScheme > -1 && uriLength > offsetAfterScheme) {
+			return uri.substring(offsetAfterScheme, uriLength);
+		}
+		return null;
+	}
+
+	public static boolean hasScheme(String uri) {
+		return getOffsetAfterScheme(uri) > -1;
+	}
+
+	public static boolean hasFileScheme(String uri) {
+
+		int offsetAfterScheme = getOffsetAfterScheme(uri);
+
+		if (offsetAfterScheme > -1) {
+			return FILE_SCHEME.equals(uri.substring(0, offsetAfterScheme));
+		}
+
+		return false;
+	}
+
+	public static int getOffsetAfterScheme(String uri) {
+		Matcher m = uriSchemePattern.matcher(uri);
+
+		if (m.matches()) {
+			return m.group(1).length();
+		}
+
+		return -1;
+	}
+
+	/**
+	 * Will return a Path object representing an existing path.
+	 * 
+	 * If this method is able to find an existing file/folder then it will return
+	 * the path to that, else it will try to find the parent directory of the
+	 * givenPath.
+	 * 
+	 * @param parentDirectory The directory that the given path is relative to
+	 * @param givenPath       Path that could be absolute or relative
+	 * @return
+	 */
+	public static Path getNormalizedPath(String parentDirectory, String givenPath) {
+
+		if (givenPath == null) {
+			return null;
+		}
+
+		int lastIndexOfSlash = givenPath.lastIndexOf(SLASH);
+
+		// in case the given path is incomplete
+		String givenPathCleaned = lastIndexOfSlash > -1 ? givenPath.substring(0, lastIndexOfSlash) : null;
+
+		Path p;
+		// The following 2 are for when the given path is already valid
+
+		p = getPathIfExists(givenPath);
+		if (p != null) {
+			// givenPath is absolute
+			return p;
+		}
+
+		p = getPathIfExists(givenPathCleaned);
+		if (p != null) {
+			// givenPath is absolute
+			return p;
+		}
+
+		if (parentDirectory == null) {
+			return null;
+		}
+
+		if (parentDirectory.endsWith(SLASH)) {
+			parentDirectory = parentDirectory.substring(0, parentDirectory.length() - 1);
+		}
+
+		String combinedPath = parentDirectory + SLASH + givenPath;
+		p = getPathIfExists(combinedPath);
+		if (p != null) {
+			return p;
+		}
+
+		combinedPath = parentDirectory + SLASH + givenPathCleaned;
+		p = getPathIfExists(combinedPath);
+		if (p != null) {
+			return p;
+		}
+
+		return null;
+	}
+
+	private static Path getPathIfExists(String path) {

```

Comment:
Simplify code like this:

private static Path getPathIfExists(String path) {
		try {
			Path p = Paths.get(path).normalize();
			return p.toFile().exists() ? p : null;
		} catch (Exception e) {
			return null;
		}
	}